salt/states/macdefaults.py

```diff
@@ -36,68 +37,101 @@ def write(name, domain, value, vtype="string", user=None):
         The value to write to the given key
 
     vtype
-        The type of value to be written, valid types are string, data, int[eger],
+        The type of value to be written. Valid types are string, data, int[eger],
         float, bool[ean], date, array, array-add, dict, dict-add
 
     user
         The user to write the defaults to
 
-
     """
     ret = {"name": name, "result": True, "comment": "", "changes": {}}
 
-    def safe_cast(val, to_type, default=None):
+    current_value = __salt__["macdefaults.read"](
+        domain, name, user, key_separator=name_separator
+    )
+
+    if vtype is not None:
         try:
-            return to_type(val)
-        except ValueError:
-            return default
-
-    current_value = __salt__["macdefaults.read"](domain, name, user)
-
-    if (vtype in ["bool", "boolean"]) and (
-        (value in [True, "TRUE", "YES"] and current_value == "1")
-        or (value in [False, "FALSE", "NO"] and current_value == "0")
-    ):
-        ret["comment"] += f"{domain} {name} is already set to {value}"
-    elif vtype in ["int", "integer"] and safe_cast(current_value, int) == safe_cast(
-        value, int
-    ):
-        ret["comment"] += f"{domain} {name} is already set to {value}"
-    elif current_value == value:
-        ret["comment"] += f"{domain} {name} is already set to {value}"
-    else:
-        out = __salt__["macdefaults.write"](domain, name, value, vtype, user)
+            value = __salt__["macdefaults.cast_value_to_vtype"](value, vtype)
+        except ValueError as e:
+            ret["result"] = False
+            ret["comment"] = f"Failed to cast value {value} to {vtype}. {e}"
+            return ret
+
+    if _compare_values(value, current_value, vtype):
+        ret["comment"] = f"{domain} {name} is already set to {value}"
+        return ret
+
+    try:
+        out = __salt__["macdefaults.write"](
+            domain, name, value, vtype, user, key_separator=name_separator
+        )
         if out["retcode"] != 0:
             ret["result"] = False
-            ret["comment"] = "Failed to write default. {}".format(out["stdout"])
+            ret["comment"] = f"Failed to write default. {out['stderr']}"
         else:
             ret["changes"]["written"] = f"{domain} {name} is set to {value}"
+    # pylint: disable-next=broad-exception-caught
+    except Exception as e:
+        ret["result"] = False
+        ret["comment"] = f"Failed to write default. {e}"
 
     return ret
 
 
-def absent(name, domain, user=None):
+def absent(name, domain, user=None, name_separator=None):
     """
     Make sure the defaults value is absent
 
     name
-        The key of the given domain to remove
+        The key of the given domain to remove.
+        It can be a nested key/index separated by `name_separator`.
+
+    name_separator
+        The separator to use when splitting the name into a list of keys.
+        If None, the name will not be split (Default).
```

Comment:
Needs `..versionadded:: 3008`

Reply:
Done in commit d9243837c283d6659df911e8c2c5a44d7ce125ff